lms/djangoapps/grades/tests/test_models.py

```diff
@@ -385,15 +385,26 @@ def test_passed_timestamp(self):
         self.assertEqual(grade.letter_grade, u'A')
         self.assertEqual(grade.passed_timestamp, passed_timestamp)
 
-        # If the grade later reverts to a failing grade, they keep their passed_timestamp
```

Comment:
I have no idea that why are they not changing passed_timestamp if users improve their scores. Seems like they did it on purpose. We can ask the relevant person about this scenario.

Reply:
I think it was done on purpose so that we can keep a record of when a learner has ever passed a course even if they later fail the course. From the [Grades Data Model page](https://openedx.atlassian.net/wiki/spaces/AC/pages/125927634/Grades+Data+Model): "passed_timestamp: Time when the learner first passed the course.  If this cell value is empty, the learner never passed the course.  If this cell value is non-empty but the letter_grade value is empty, the learner transitioned from passing to not passing."

Unless we have a good reason for it, I think we should keep the behavior that the passed_timestamp remains the same even after the learner reverts to a failing grade. It sounds like we are relying on that behavior for measuring certain metrics.

Reply:
@thallada yes this is also one of the metric from the other metrics due to which we were still able to see `Download Now` link even after user failed in the course.
https://github.com/edx/edx-platform/blob/dd93e67b5d776f4178b5243d78c74a88e05a39ce/lms/djangoapps/grades/course_grade_factory.py#L150-L156

So, we need to set passed_timestamp to None in order to run the things correctly.

Reply:
I checked with @nasthagiri. She thinks that instead of overwriting `passed_timestamp` (which is supposed to be a permanent historical record of the first time a learner ever passed), we should fix the code that you referenced so that it checks the `letter_grade` instead of `passed_timestamp` to determine passing status:

```python
 return CourseGrade( 
     user, 
     course_data, 
     persistent_grade.percent_grade, 
     persistent_grade.letter_grade, 
     persistent_grade.letter_grade is not None, 
 )
```

Reply:
I have changed the code with your proposed fix.